File: flink-app/src/test/java/com/example/ctr/domain/model/EventCountTest.java
package com.example.ctr.domain.model;

import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThat;

class EventCountTest {

    @Test
    void initial_ShouldReturnZeroCount() {
        // When
        EventCount eventCount = EventCount.initial();

        // Then
        assertThat(eventCount.getImpressions()).isEqualTo(0L);
        assertThat(eventCount.getClicks()).isEqualTo(0L);
    }

    @Test
    void incrementImpressions_ShouldIncreaseImpressionsCount() {
        // Given
        EventCount eventCount = EventCount.initial();

        // When
        eventCount.incrementImpressions();
        eventCount.incrementImpressions();

        // Then
        assertThat(eventCount.getImpressions()).isEqualTo(2L);
        assertThat(eventCount.getClicks()).isEqualTo(0L);
    }

    @Test
    void incrementClicks_ShouldIncreaseClicksCount() {
        // Given
        EventCount eventCount = EventCount.initial();

        // When
        eventCount.incrementClicks();
        eventCount.incrementClicks();
        eventCount.incrementClicks();

        // Then
        assertThat(eventCount.getImpressions()).isEqualTo(0L);
        assertThat(eventCount.getClicks()).isEqualTo(3L);
    }

    @Test
    void merge_ShouldCombineTwoEventCounts() {
        // Given
        EventCount count1 = new EventCount(100L, 5L);
        EventCount count2 = new EventCount(200L, 10L);

        // When
        EventCount merged = EventCount.merge(count1, count2);

        // Then
        assertThat(merged.getImpressions()).isEqualTo(300L);
        assertThat(merged.getClicks()).isEqualTo(15L);
    }

    @Test
    void copy_ShouldCreateIndependentCopy() {
        // Given
        EventCount original = new EventCount(100L, 5L);

        // When
        EventCount copy = new EventCount(original.getImpressions(), original.getClicks());
        copy.incrementImpressions();
        copy.incrementClicks();

        // Then
        assertThat(original.getImpressions()).isEqualTo(100L);
        assertThat(original.getClicks()).isEqualTo(5L);
        assertThat(copy.getImpressions()).isEqualTo(101L);
        assertThat(copy.getClicks()).isEqualTo(6L);
    }

    @Test
    void mixedOperations_ShouldWorkCorrectly() {
        // Given
        EventCount eventCount = EventCount.initial();

        // When
        eventCount.incrementImpressions();
        eventCount.incrementImpressions();
        eventCount.incrementClicks();

        // Then
        assertThat(eventCount.getImpressions()).isEqualTo(2L);
        assertThat(eventCount.getClicks()).isEqualTo(1L);
    }
}

File: flink-app/src/test/kotlin/com/example/ctr/domain/model/EventCountTest.kt
package com.example.ctr.domain.model

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class EventCountTest {

    @Test
    fun `initial should return zero count`() {
        val eventCount = EventCount.initial()
        assertThat(eventCount.impressions).isEqualTo(0L)
        assertThat(eventCount.clicks).isEqualTo(0L)
    }

    @Test
    fun `incrementImpressions should increase value`() {
        val eventCount = EventCount.initial()
        eventCount.incrementImpressions()
        eventCount.incrementImpressions()
        assertThat(eventCount.impressions).isEqualTo(2L)
        assertThat(eventCount.clicks).isZero()
    }

    @Test
    fun `incrementClicks should increase value`() {
        val eventCount = EventCount.initial()
        repeat(3) { eventCount.incrementClicks() }
        assertThat(eventCount.clicks).isEqualTo(3L)
        assertThat(eventCount.impressions).isZero()
    }

    @Test
    fun `merge should combine two counts`() {
        val count1 = EventCount(100L, 5L)
        val count2 = EventCount(200L, 10L)

        val merged = EventCount.merge(count1, count2)
        assertThat(merged.impressions).isEqualTo(300L)
        assertThat(merged.clicks).isEqualTo(15L)
    }

    @Test
    fun `copy should create independent copy`() {
        val original = EventCount(100L, 5L)
        val copy = original.copy()
        copy.incrementImpressions()
        copy.incrementClicks()

        assertThat(original.impressions).isEqualTo(100L)
        assertThat(original.clicks).isEqualTo(5L)
        assertThat(copy.impressions).isEqualTo(101L)
        assertThat(copy.clicks).isEqualTo(6L)
    }

    @Test
    fun `mixed operations should work`() {
        val eventCount = EventCount.initial()
        eventCount.incrementImpressions()
        eventCount.incrementImpressions()
        eventCount.incrementClicks()

        assertThat(eventCount.impressions).isEqualTo(2L)
        assertThat(eventCount.clicks).isEqualTo(1L)
    }
}
